### PR TITLE
Ubuntu 24.10: port missing slices from 24.04

### LINKED ISSUE #391: Port forward recent 24.04 additions to 24.10 branch
The ubuntu-24.10 branch has been available and ready for a couple of weeks now, but a couple of PRs targeting the ubuntu-24.04 branch have been merged  since and have not been port forwarded, leading to some SDFs being available in 24.04 and not in 24.10.
Over time this will cause issues, see [CONTRIBUTING.md](https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md#2-which-ubuntu-releases-are-you-targeting):
> To enforce consistency across the different chisel releases, every slice definition that is proposed for a given chisel release MUST also be proposed for all newer and still supported chisel releases

The goal of this issue is to track all PRs that need to be port-forwarded to the 24.10 branch. Those are as follows (in order of merge to the 24.04 branch):
| 24.04 PR | 24.10 PR |
|---|---|
| #364 | TODO |
| #220 | TODO |
| #350 | TODO |
| #349 | TODO |
| #360 | TODO |
| #375 | #376 |
| #370 | #371 |
| #341 | TODO |
| #324 | TODO |
| #359 | TODO |
| #314 | TODO |

### PR DESCRIPTION
**Pull request by cjdcordeiro**
# Proposed changes
There were some standing 24.04 PRs at the time of opening 24.10 that weren't forward-ported. This PR ports them from 24.04.

## Related issues/PRs
Closes #391 

## Checklist

* [x] I have read the [contributing guidelines](
https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md)
* [x] I have tested my changes ([see how](https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md#7-test-your-slices-before-opening-a-pr))
* [x] I have already submitted the [CLA form](
https://ubuntu.com/legal/contributors/agreement)

### CODE REVIEW

**Comment by vpa1977** [comment]
> @vpa1977 could you please help understand the java errors in these tests: https://github.com/canonical/chisel-releases/actions/runs/11915704026/job/33206594753?pr=392
> 
> I can't reproduce them locally. I just copied them from 24.04.

It looks like spread is reusing the vms? 
```
VM Started: + chroot . usr/lib/jvm/java-11-openjdk-amd64/bin/jcmd MonitoringTest VM.version

[1045](https://github.com/canonical/chisel-releases/actions/runs/11915704026/job/33206594753?pr=392#step:9:1046)10636:

[1046](https://github.com/canonical/chisel-releases/actions/runs/11915704026/job/33206594753?pr=392#step:9:1047)OpenJDK 64-Bit Server VM version 17.0.13+11-Ubuntu-2ubuntu124.10
```
We should not have java 17 process, but yet it is present. The tests expect the same/lower version Java to be installed.

**Comment by cjdcordeiro** [comment]
@vpa1977 indeed since this PR is running many tests in parallel, multiple versions of java get installed at the same time and may clash.

I've adjusted the java tests in order to reduce the changes of such clashes by:
 - cleaning up after the task execution
 - enforcing the use of the right java alternative at test time

**Comment by vpa1977** [comment]
> @vpa1977 indeed since this PR is running many tests in parallel, multiple versions of java get installed at the same time and may clash.
> 
> I've adjusted the java tests in order to reduce the changes of such clashes by:
There is a typo/not critical thing in Java 17 tests - it uses Java 11 on the host, hence the test failure. 
We can either bump host to 17 or use 11 update-alternatives.

**Comment by cjdcordeiro** [comment]
thanks for pointing that out @vpa1977

File: tests/spread/integration/openjdk-11-jre-headless/task.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      tools)
+        DNAME="CN=Sample Cert, OU=R&D, O=Company Ltd., L=Dublin 4, S=Dublin, C=IE"
+        chroot . $(dirname ${java})/keytool -genkeypair -keystore foo -storepass barbar -keyalg RSA -dname "$DNAME"
+        chroot . $(dirname ${java})/pack200 foo.pack.gz test.jar
+        chroot . $(dirname ${java})/unpack200 foo.pack.gz test1.jar
+      ;;
+    esac
+  done
+
+restore: apt remove -y openjdk-11-jre-headless
```
Comment (rebornplusplus) [inline tests/spread/integration/openjdk-11-jre-headless/task.yaml:67]:
```suggestion
restore: apt remove -y openjdk-11-jdk-headless
```

Comment (rebornplusplus) [reply tests/spread/integration/openjdk-11-jre-headless/task.yaml:67]: ^ We are installing the `jdk` package above, not the `jre`.

File: tests/spread/integration/openjdk-17-jre-headless/task.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        chroot . ${java} -cp . ReadCertificate
+      ;;
+      tools)
+        DNAME="CN=Sample Cert, OU=R&D, O=Company Ltd., L=Dublin 4, S=Dublin, C=IE"
+        chroot . $(dirname ${java})/keytool -genkeypair -keystore foo -storepass barbar -keyalg RSA -dname "$DNAME"
+      ;;
+    esac
+  done
+
+restore: apt remove -y openjdk-11-jre-headless
```
Comment (rebornplusplus) [inline tests/spread/integration/openjdk-17-jre-headless/task.yaml:66]:
```suggestion
restore: apt remove -y openjdk-11-jdk-headless
```

We are installing the `jdk` package, not the `jre`.

**Review by rebornplusplus — CHANGES_REQUESTED** [review]
Quick review of the java spread fixes.

File: tests/spread/integration/openjdk-11-jre-headless/task.yaml (hunk 1 above)
Comment (cjdcordeiro) [reply tests/spread/integration/openjdk-11-jre-headless/task.yaml:67]: good catch ty

File: tests/spread/integration/openjdk-17-jre-headless/task.yaml (hunk 2 above)
Comment (cjdcordeiro) [reply tests/spread/integration/openjdk-17-jre-headless/task.yaml:66]: ty

File: tests/spread/integration/openjdk-11-jre-headless/task.yaml (hunk 1 above)
Comment (vpa1977) [reply tests/spread/integration/openjdk-11-jre-headless/task.yaml]: Actually first option was correct. Removing JRE will remove both jdk and JRE. To remove all of installed java versions we can remove ca-certificates-java.

Removing JDK works too, but it leaves JRE  installed.

**Review by vpa1977 — APPROVED** [review]
Scope of review: Java slices and tests.

**Comment by norrisjeremy** [comment]
When will this be merged?

**Comment by clay-lake** [comment]
Hi @norrisjeremy, I'm currently reviewing this. We should have an update soon.

**Review by clay-lake — APPROVED** [review]
This should be good to go. I looked at differences between releases and dependencies I could not spot anything that stood out. @cjdcordeiro 

@rebornplusplus Do you want to update your status here?

**Pull request merged**